### PR TITLE
Remove `PSv2CompletionCompleter`

### PR DESCRIPTION
**Pull request by xtqqczze**
Contributes to #18323.

Remove references to "PSv2" in: `src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs`

_Breaking changes_ to public APIs:
* [CompleteInput(String, Int32, Hashtable, PowerShell)](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.commandcompletion.completeinput?view=powershellsdk-7.0.0#system-management-automation-commandcompletion-completeinput(system-string-system-int32-system-collections-hashtable-system-management-automation-powershell))
* [CompleteInput(Ast, Token[], IScriptPosition, Hashtable, PowerShell)](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.commandcompletion.completeinput?view=powershellsdk-7.0.0#system-management-automation-commandcompletion-completeinput(system-management-automation-language-ast-system-management-automation-language-token()-system-management-automation-language-iscriptposition-system-collections-hashtable-system-management-automation-powershell))

### CODE REVIEW

**Comment by xtqqczze** [comment]
`LastWordFinder` is also legacy (used by `InvokeLegacyTabExpansion`) and could be removed:
https://github.com/PowerShell/PowerShell/blob/920bf3b80dd5236441ac3ab3bf2ca644baa4c625/src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs#L670-L673

**Comment by xtqqczze** [comment]
e86ca2a: This commit makes changes to public APIs:
* [CompleteInput(String, Int32, Hashtable, PowerShell)](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.commandcompletion.completeinput?view=powershellsdk-7.0.0#system-management-automation-commandcompletion-completeinput(system-string-system-int32-system-collections-hashtable-system-management-automation-powershell))
* [CompleteInput(Ast, Token[], IScriptPosition, Hashtable, PowerShell)](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.commandcompletion.completeinput?view=powershellsdk-7.0.0#system-management-automation-commandcompletion-completeinput(system-management-automation-language-ast-system-management-automation-language-token()-system-management-automation-language-iscriptposition-system-collections-hashtable-system-management-automation-powershell))

However, it seems the legacy `TabExpansion` internal function (as opposed to `TabExpansion2`) is a PSv2 specific so maybe we can drop support.

**Comment by xtqqczze** [comment]
Ready for review. I wasn't sure how to resolve failures in e86ca2a so partially reverted, this PR is already large enough.

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 1]
```diff
@@ -178,17 +178,11 @@ public static CommandCompletion CompleteInput(string input, int cursorIndex, Has
...
-                        // Remoting to a Win7 machine. Use the legacy tab completion function from V1/V2
-                        int replacementIndex;
-                        int replacementLength;
-
-                        powershell.Commands.Clear();
-                        var results = InvokeLegacyTabExpansion(powershell, input, cursorIndex, true, out replacementIndex, out replacementLength);
-                        return new CommandCompletion(
-                            new Collection<CompletionResult>(results ?? EmptyCompletionResult),
-                            -1, replacementIndex, replacementLength);
+                        throw new NotSupportedException("Remote server must be PSv3 or later.");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:185]:
```c#
// Remoting to a Win7 machine.
return s_emptyCommandCompletion;
```

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 2]
```diff
@@ -178,17 +178,11 @@ public static CommandCompletion CompleteInput(string input, int cursorIndex, Has
                 if (!powershell.IsChild)
                 {
                     CheckScriptCallOnRemoteRunspace(remoteRunspace);
+
+                    // If the runspace has only legacy capabilities, the remote server is older than PSv3.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:182]: Please reword or remove the comment.

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 3]
```diff
@@ -249,22 +243,11 @@ public static CommandCompletion CompleteInput(Ast ast, Token[] tokens, IScriptPo
...
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:247]: The same.

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 4]
```diff
@@ -249,22 +243,11 @@ public static CommandCompletion CompleteInput(Ast ast, Token[] tokens, IScriptPo
...
-                        int replacementLength;
-
-                        // When call the win7 TabExpansion script, the input should be the single current line
-                        powershell.Commands.Clear();
-                        var inputAndCursor = GetInputAndCursorFromAst(cursorPosition);
-                        var results = InvokeLegacyTabExpansion(powershell, inputAndCursor.Item1, inputAndCursor.Item2, true, out replacementIndex, out replacementLength);
-                        return new CommandCompletion(
-                            new Collection<CompletionResult>(results ?? EmptyCompletionResult),
-                            -1, replacementIndex + inputAndCursor.Item3, replacementLength);
+                        throw new NotSupportedException("Remote server must be running on PSv3 or later.");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:250]: The same.
```c#
// Remoting to a Win7 machine.
return s_emptyCommandCompletion;
```

**Comment by xtqqczze** [comment]
I tried to remove this code in e86ca2ae0106f85462a8af2b8438fabe83589a0a, but I could not resolve the test failures.

https://github.com/PowerShell/PowerShell/blob/5996d49b58e6cfc0abd21ba5fc2d034aab8099ac/src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs#L504-L515

**Comment by xtqqczze** [comment]
PowerShell-CI-windows (Test for Windows Windows Test - ElevatedPesterTests - Others)  failure:
> Set/New/Remove-Service cmdlet tests.Remove-Service can accept a ServiceController as pipeline input

**Comment by xtqqczze** [comment]
restarting tests

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 5]
```diff
@@ -180,17 +177,11 @@ public static CommandCompletion CompleteInput(string input, int cursorIndex, Has
...
-                        // Remoting to a Win7 machine. Use the legacy tab completion function from V1/V2
-                        int replacementIndex;
-                        int replacementLength;
-
-                        powershell.Commands.Clear();
-                        var results = InvokeLegacyTabExpansion(powershell, input, cursorIndex, true, out replacementIndex, out replacementLength);
-                        return new CommandCompletion(
-                            new Collection<CompletionResult>(results ?? EmptyCompletionResult),
-                            -1, replacementIndex, replacementLength);
+                        // Remoting to a Win7 machine.
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:183]: Technically a Win7 machine can have 5.1 installed via WMF5.1, so this comment should really be remoting to a PS2.0 machine

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 6]
```diff
@@ -251,22 +241,11 @@ public static CommandCompletion CompleteInput(Ast ast, Token[] tokens, IScriptPo
...
-                        int replacementLength;
-
-                        // When call the win7 TabExpansion script, the input should be the single current line
-                        powershell.Commands.Clear();
-                        var inputAndCursor = GetInputAndCursorFromAst(cursorPosition);
-                        var results = InvokeLegacyTabExpansion(powershell, inputAndCursor.Item1, inputAndCursor.Item2, true, out replacementIndex, out replacementLength);
-                        return new CommandCompletion(
-                            new Collection<CompletionResult>(results ?? EmptyCompletionResult),
-                            -1, replacementIndex + inputAndCursor.Item3, replacementLength);
+                        // Remoting to a Win7 machine.
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:247]: Same here

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 7]
```diff
@@ -625,7 +604,7 @@ private static bool NeedToInvokeLegacyTabExpansion(PowerShell powershell)
             return false;
         }
 
-        private static List<CompletionResult> InvokeLegacyTabExpansion(PowerShell powershell, string input, int cursorIndex, bool remoteToWin7, out int replacementIndex, out int replacementLength)
+        private static List<CompletionResult> InvokeLegacyTabExpansion(PowerShell powershell, string input, int cursorIndex, out int replacementIndex, out int replacementLength)
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:607]: This method should just be removed as it seems all calls are within this source file

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.LastWordFinder.cs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+namespace System.Management.Automation
+{
+    public partial class CommandCompletion
+    {
+        /// <summary>
+        /// LastWordFinder implements the algorithm we use to search for the last word in a line of input taken from the console.
+        /// This class exists for legacy purposes only - V3 and forward uses a slightly different interface.
+        /// </summary>
+        private sealed class LastWordFinder
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.LastWordFinder.cs:12]: The only use of this seems to be for legacy and should be removed

Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.LastWordFinder.cs:12]: I cloned your branch and removed the legacy code and builds fine, we'll have to look at the test failures (if any) with those changes

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
some legacy code should be removed

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.LastWordFinder.cs (hunk 8 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.LastWordFinder.cs]: Test failures in [build 111723](https://dev.azure.com/powershell/2972bb5c-f20c-4a60-8bd9-00ffe9987edc/_build/results?buildId=111723) (e86ca2ae0106f85462a8af2b8438fabe83589a0a):
```log
 TabCompletion.User-overridden TabExpansion implementations.Override TabExpansion with function

Exception calling "CompleteInput" with "3" argument(s): "Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')"
```

```log
 TabCompletion.User-overridden TabExpansion implementations.Override TabExpansion with alias

Exception calling "CompleteInput" with "3" argument(s): "Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')"
```

I've reset to e86ca2ae0106f85462a8af2b8438fabe83589a0a and applied changes from review.

**Comment by xtqqczze** [comment]
Removed "Override TabExpansion with function" and "Override TabExpansion with alias" tests as they relate to the removed code.

**Comment by xtqqczze** [comment]
restarting tests

**Comment by xtqqczze** [comment]
@SteveL-MSFT Ready for review, feedback addressed and test failures resolved.

File: src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs [hunk 9]
```diff
@@ -584,737 +550,21 @@ private static CommandCompletion CompleteInputImpl(Ast ast, Token[] tokens, IScr
                     }
                     */
                 }
-
+
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:553]: Please remove.

Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs:553]:
```suggestion

```

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Pull request merged**